src/main/java/actionlogger/trackers/InventoryTracker.java

```diff
@@ -4,42 +4,69 @@
 import lombok.RequiredArgsConstructor;
 import lombok.Value;
 import lombok.extern.slf4j.Slf4j;
-import net.runelite.api.InventoryID;
 import net.runelite.api.Item;
 import net.runelite.api.events.ItemContainerChanged;
+import net.runelite.api.gameval.InventoryID;
 import net.runelite.client.eventbus.Subscribe;
 
 import java.util.Arrays;
-import java.util.List;
-import java.util.stream.Collectors;
 
 @Slf4j
 @RequiredArgsConstructor
 public class InventoryTracker {
     private static final String TYPE = "INVENTORY_CHANGED";
+    private static final int INVENTORY_SIZE = 28;
 
     private final JsonWriter writer;
 
-    private List<Integer> inventoryItemIDs = List.of(-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1);
+    private int[] inventoryItemIDs = filled(INVENTORY_SIZE, -1);
+    private int[] inventoryQuantities = filled(INVENTORY_SIZE, 0);
 
     @Subscribe
     public void onItemContainerChanged(ItemContainerChanged event) {
-        if (event.getContainerId() != InventoryID.INVENTORY.getId()) {
+        if (event.getContainerId() != InventoryID.INV) {
             return;
         }
 
         var items = event.getItemContainer().getItems();
-        var itemIDs = Arrays.stream(items).map(Item::getId).collect(Collectors.toList());
+        var itemIDs = Arrays.stream(items).mapToInt(Item::getId).toArray();
+        var itemQuantities = Arrays.stream(items).mapToInt(Item::getQuantity).toArray();
 
-        if (!inventoryItemIDs.equals(itemIDs)) {
-            this.writer.write(TYPE, new InventoryChangedData(this.inventoryItemIDs, itemIDs));
+        boolean idChanged = inventoryItemIDs.length != itemIDs.length;
+        boolean quantityChanged = false;
+        final int n = Math.min(inventoryItemIDs.length, itemIDs.length);
+        for (int i = 0; i < n; i++) {
+            if (inventoryItemIDs[i] != itemIDs[i]) {
+                idChanged = true;
+            } else if (inventoryQuantities[i] != itemQuantities[i]) {
+                quantityChanged = true;
+            }
+        }
+
+        if (idChanged || quantityChanged) {
+            var data = new InventoryChangedData(
+                idChanged ? this.inventoryItemIDs : null,
```

Comment:
`oldInventory` is not populated if only quantities changed, while item IDs were constant

are we concerned that an existing parser impl could fail if it always expected this field to be present?

Reply:
the other benefit is it allowed $O(1)$ check of whether id or quantity changed (now need to compare full arrays)

Reply:
Yeah I think just include all data, and if we need it we just add bools for what actually changed in the future